### PR TITLE
fix(operator): derive readiness from child resources

### LINKED ISSUE #79: fix(operator): derive Ready status from actual resource readiness
## Summary

The operator reports resources as Ready before the backing Kubernetes controllers have made them usable. BrewFSCluster becomes Ready immediately after API objects are applied, without checking Redis/RustFS Pods, the PVC, or the bucket-init Job. A newly created DaemonSet mount/consumer can also be marked Ready because missing status values are compared as 0 >= 0.

Reviewed at commit 25ca7fce9cc49eefca07ef8024f95c13e7a595dc.

## Evidence

Cluster reconciliation applies resources and immediately patches status:

- https://github.com/brewfs/brewfs/blob/25ca7fce9cc49eefca07ef8024f95c13e7a595dc/operator/brewfs-operator/src/reconciler.rs#L58-L72

cluster_ready_status is unconditional and does not inspect any child-resource status:

- https://github.com/brewfs/brewfs/blob/25ca7fce9cc49eefca07ef8024f95c13e7a595dc/operator/brewfs-operator/src/reconciler.rs#L1860-L1874

Mount reconciliation checks only that the cluster object and ConfigMap exist, not that the cluster is actually ready:

- https://github.com/brewfs/brewfs/blob/25ca7fce9cc49eefca07ef8024f95c13e7a595dc/operator/brewfs-operator/src/reconciler.rs#L90-L170

The readiness comparison defaults both absent values to zero:

- https://github.com/brewfs/brewfs/blob/25ca7fce9cc49eefca07ef8024f95c13e7a595dc/operator/brewfs-operator/src/reconciler.rs#L181-L187

For DaemonSets, both desired and ready values come from status, which may be absent immediately after creation:

- https://github.com/brewfs/brewfs/blob/25ca7fce9cc49eefca07ef8024f95c13e7a595dc/operator/brewfs-operator/src/reconciler.rs#L1683-L1714

## Impact

Automation and users can observe Ready while the PVC is pending, backends are unavailable, bucket creation failed/is still running, or zero mount Pods have been scheduled. Dependent consumer workloads can start against an unusable mount.

## Suggested fix

Derive status conditions from observed child resources and observedGeneration. Require the PVC to be Bound, Redis/RustFS rollouts to be available, and the init Job to have completed for the current effective spec. Treat missing workload/DaemonSet status as Progressing, not zero-of-zero Ready. Gate mount reconciliation on the cluster Ready condition.

Add reconciliation tests for newly created resources, pending PVC, failed Job, unavailable Deployment, empty DaemonSet status, and a successful transition to Ready.

### PR DESCRIPTION
**Pull request by marshawcoco**
## Summary
- derive BrewFSCluster readiness from current Redis/RustFS Deployment rollouts, the RustFS PVC, and the bucket-init Job
- watch owned backend resources and referenced BrewFSMount cluster changes so status and mounts react promptly
- gate mount and consumer startup on a Ready cluster, clean up stale workloads while progressing, and prevent consumers from running against an unavailable mount
- enforce current-generation, rollout-complete readiness with kind-specific replica semantics, restart failed init Jobs, and add Redis command and RustFS HTTP readiness probes
- document readiness semantics and add focused status regressions

## Why
The operator previously marked a cluster Ready immediately after applying child objects and treated stale or missing workload counters as sufficient. Mount and consumer workloads could continue running through backend outages, image or kind changes, or failed bucket initialization, while child changes could leave cluster status stale.

## Validation
- `cargo test --quiet` (operator): 10 passed
- `CARGO_NET_OFFLINE=true CARGO_BUILD_JOBS=1 cargo test --quiet --all-features` (operator): 20 passed
- `CARGO_NET_OFFLINE=true CARGO_BUILD_JOBS=1 cargo clippy --quiet --all-targets --all-features -- -D warnings`
- `cargo fmt --all --check`
- `git diff --check`

## Risk
Clusters remain Progressing until child controllers publish current rollout readiness, the RustFS PVC is Bound, and bucket initialization succeeds. Mount and consumer workloads may be recreated after backend recovery or spec changes, which prevents stale processes and premature consumer startup.

Fixes #79

### CODE REVIEW

File: operator/brewfs-operator/src/reconciler.rs [hunk 1]
```diff
@@ -143,6 +145,38 @@ pub async fn reconcile_mount(
         .and_then(|status| status.config_map.clone())
         .unwrap_or_else(|| cluster_config_map_name(&cluster.name_any()));
 
+    if !cluster_status_is_ready(&cluster) {
```
Comment (Ivanbeethoven) [inline operator/brewfs-operator/src/reconciler.rs:148]: **Do not skip cleanup while the cluster is Progressing.**

This early return also skips all mount/consumer apply and delete paths. If an existing mount removes `consumer`, switches workload kind, or changes its image while the cluster remains non-Ready, the old consumer/workload keeps running indefinitely because it is owned by the still-existing `BrewFSMount`. Please gate creation/startup on backend readiness without bypassing destructive reconciliation and stale-resource cleanup.

File: operator/brewfs-operator/src/reconciler.rs [hunk 2]
```diff
@@ -1868,15 +1908,156 @@ layout:\n  chunk_size: {chunk_size}\n  block_size: {block_size}\n",
...
+        .get_opt(&job_name)
+        .await
+        .with_context(|| format!("load RustFS init Job {job_name}"))?
+    else {
+        return Ok((
+            "Progressing".to_string(),
+            format!("waiting for RustFS init Job {job_name}"),
+        ));
+    };
+    if !job_is_succeeded(&job) {
```
Comment (Ivanbeethoven) [inline operator/brewfs-operator/src/reconciler.rs:2202]: **Recover from a terminally failed init Job.**

If this fixed-name Job exhausts `backoffLimit` during a temporary RustFS outage, later SSA reconciles keep the same terminal Job and never rerun it. Even after RustFS recovers, this check therefore leaves the cluster and every mount blocked until someone manually deletes the Job. The reconciler needs to replace/recreate failed attempts (or otherwise make retries restartable), not only observe `succeeded`.

File: operator/brewfs-operator/src/reconciler.rs [hunk 3]
```diff
@@ -191,7 +225,7 @@ pub async fn reconcile_mount(
     .await?;
     let (desired_replicas, ready_replicas) =
         get_mount_workload_status(&client, &namespace, &mount, &workload_name).await?;
-    let phase = if ready_replicas.unwrap_or_default() >= desired_replicas.unwrap_or_default() {
+    let phase = if replicas_are_ready(desired_replicas, ready_replicas) {
```
Comment (Ivanbeethoven) [inline operator/brewfs-operator/src/reconciler.rs]: **Require status from the current workload revision.**

Immediately after applying an image/template change, these counters can still describe the previous workload revision. For example, old `desired=1, ready=1` can cause the new `BrewFSMount` generation to be patched `Ready` even when the requested image has no ready pod. Deployment/DaemonSet/StatefulSet readiness should also validate observed generation and the kind-specific updated/current revision fields.

File: operator/brewfs-operator/src/reconciler.rs [hunk 4]
```diff
@@ -1868,15 +1908,156 @@ layout:\n  chunk_size: {chunk_size}\n  block_size: {block_size}\n",
...
+        return false;
+    }
+    if let Some(generation) = deployment.metadata.generation {
+        if status.observed_generation < Some(generation) {
+            return false;
+        }
+    }
+    status.ready_replicas.unwrap_or_default() >= desired
+        && status.available_replicas.unwrap_or_default() >= desired
+        && status.updated_replicas.unwrap_or_default() >= desired
```
Comment (Ivanbeethoven) [inline operator/brewfs-operator/src/reconciler.rs]: **Do not combine readiness counters from different revisions.**

During a one-replica rolling update, the old pod can supply `readyReplicas=1` and `availableReplicas=1` while the surged, unready new pod supplies `updatedReplicas=1`. All three independent comparisons then pass even though no updated pod is ready. Please use a rollout-complete predicate that also excludes old replicas (for example current observed generation plus updated/ready/available counts and no unavailable/old replicas).

File: operator/brewfs-operator/src/reconciler.rs [hunk 5]
```diff
@@ -75,12 +75,14 @@ pub async fn reconcile_cluster(
...
     apply_brewfs_config(&client, &namespace, &cluster, &owner).await?;
-    patch_cluster_status(&client, &namespace, &cluster).await?;
+    let (phase, message) = observe_cluster_readiness(&client, &namespace, &cluster).await?;
+    patch_cluster_status(&client, &namespace, &cluster, &phase, &message).await?;
     #[cfg(feature = "workspace-operator")]
     crate::workspace::controller::reconcile_cluster_workspace(&cluster, &client, &namespace)
         .await?;
 
-    Ok(Action::requeue(Duration::from_secs(300)))
+    let requeue_after = if phase == "Ready" { 300 } else { 15 };
```
Comment (Ivanbeethoven) [inline operator/brewfs-operator/src/reconciler.rs:84]: **Watch owned child resources instead of leaving Ready stale.**

The cluster controller watches only `BrewFSCluster`; it does not register its Deployments, PVC, or Job. If a backend loses availability just after a Ready reconcile, no child event updates status, so mounts can pass the stale Ready gate for up to this 300-second poll interval. Register the owned resources with the controller (and watch referenced cluster changes for mounts), keeping a long requeue only as a safety net.

File: operator/brewfs-operator/src/reconciler.rs [hunk 6]
```diff
@@ -1868,15 +1908,156 @@ layout:\n  chunk_size: {chunk_size}\n  block_size: {block_size}\n",
...
+        .get_opt(&redis_name)
+        .await
+        .with_context(|| format!("load Redis Deployment {redis_name}"))?
+    else {
+        return Ok((
+            "Progressing".to_string(),
+            format!("waiting for Redis Deployment {redis_name}"),
+        ));
+    };
+    if !deployment_is_ready(&redis) {
```
Comment (Ivanbeethoven) [inline operator/brewfs-operator/src/reconciler.rs:2146]: **Make backend pod readiness reflect service readiness.**

The generated Redis and RustFS containers have no readiness/startup probe, so Deployment `readyReplicas` only proves that the containers are Running. Redis can still be loading AOF, and a restarted RustFS pod can be unavailable after the already-successful init Job, yet this gate marks the cluster Ready. Add service-level probes before deriving backend readiness from Deployment status.

File: operator/brewfs-operator/src/reconciler.rs [hunk 7]
```diff
@@ -1691,6 +1725,12 @@ fn mount_propagation_value(mode: &MountPropagationMode) -> &'static str {
     }
 }
 
+fn replicas_are_ready(desired: Option<i32>, ready: Option<i32>) -> bool {
+    desired
+        .zip(ready)
+        .is_some_and(|(desired, ready)| ready >= desired)
```
Comment (Ivanbeethoven) [inline operator/brewfs-operator/src/reconciler.rs]: **Handle zero desired replicas per workload kind.**

This shared predicate gives contradictory zero-count results: a scaled-to-zero Deployment/StatefulSet commonly has `(Some(0), None)` and stays `Progressing`, while a DaemonSet with no eligible nodes has `(Some(0), Some(0))` and is reported `Ready` despite no mount pod. The readiness policy needs kind-specific zero semantics rather than only comparing two optional counts.

File: operator/brewfs-operator/src/reconciler.rs [hunk 8]
```diff
@@ -230,7 +264,7 @@ pub async fn reconcile_mount(
             consumer_desired_replicas = desired;
             consumer_ready_replicas = ready;
 
-            if ready.unwrap_or_default() < desired.unwrap_or_default() {
+            if !replicas_are_ready(desired, ready) {
```
Comment (Ivanbeethoven) [inline operator/brewfs-operator/src/reconciler.rs]: **Wait for the mount before creating its consumer.**

Before reaching this readiness check, the branch creates the consumer even when the mount workload itself is still `Progressing`. The generated `DirectoryOrCreate` hostPath can therefore let the consumer start against an empty directory; there is also no generated wait or same-node constraint. Please create/enable the consumer only after the mount is usable on the consumer's node.

File: operator/brewfs-operator/src/reconciler.rs [hunk 9]
```diff
@@ -1868,15 +2037,211 @@ layout:\n  chunk_size: {chunk_size}\n  block_size: {block_size}\n",
...
+            format!("waiting for RustFS init Job {job_name}"),
+        ));
+    };
+    if !job_is_succeeded(&job) {
+        let failed = job
+            .status
+            .as_ref()
+            .and_then(|status| status.failed)
+            .unwrap_or_default();
+        let message = if failed > 0 {
```
Comment (Ivanbeethoven) [inline operator/brewfs-operator/src/reconciler.rs:2599]: **Do not treat any failed Pod as a terminally failed Job.**

`status.failed > 0` is not equivalent to the Job being terminally Failed. With `backoffLimit > 1`, Kubernetes may have one or more failed Pods while the Job controller is still retrying. Deleting/recreating here resets that retry/backoff state and can turn a transient failure into an endless restart loop.

Please key retry/recreation off the terminal Job condition (`type=Failed,status=True`) rather than the raw cumulative failed-Pod count. Add a regression where `failed=1`, `active=1` (or otherwise non-terminal) and verify the Job is left in place.

Comment (Ivanbeethoven) [inline operator/brewfs-operator/src/reconciler.rs:2599]: **Do not treat any failed Pod as a terminally failed Job.**

`status.failed > 0` is not equivalent to the Job being terminally Failed. With `backoffLimit > 1`, Kubernetes may have one or more failed Pods while the Job controller is still retrying. Deleting/recreating here resets that retry/backoff state and can turn a transient failure into an endless restart loop.

Please key retry/recreation off the terminal Job condition (`type=Failed,status=True`) rather than the raw cumulative failed-Pod count. Add a regression where `failed=1`, `active=1` and the Job is not terminal; it should be left in place.

File: operator/brewfs-operator/src/reconciler.rs [hunk 10]
```diff
@@ -216,28 +264,55 @@ pub async fn reconcile_mount(
...
-            .await?;
-            let (desired, ready) =
-                get_consumer_workload_status(&client, &namespace, consumer, &consumer_name).await?;
-            consumer_workload_name_status = Some(consumer_name.clone());
-            consumer_desired_replicas = desired;
-            consumer_ready_replicas = ready;
-
-            if ready.unwrap_or_default() < desired.unwrap_or_default() {
+            let consumer_name = consumer_workload_name(&mount.name_any());
+            if phase != "Ready" {
```
Comment (Ivanbeethoven) [inline operator/brewfs-operator/src/reconciler.rs:320]: **Global mount readiness is not enough for a node-local `hostPath`.**

Waiting until the mount workload is Ready fixes startup ordering, but the consumer can still be scheduled onto a node where the BrewFS mount is absent. `DirectoryOrCreate` then creates an empty directory and the consumer starts against the wrong filesystem.

This is reachable with a one-replica mount Deployment on node A and a consumer Deployment on node B; a DaemonSet also needs to cover every node eligible for the consumer. Please make node-local mount availability part of the scheduling contract (compatible affinity, a mount DaemonSet on all eligible nodes, or a per-node readiness label/marker) and add a multi-node regression.
